File: src/vector_calculations.jl
module Calculations
	using ..StateContainers: State, check_state
	export recalculate_variables!, recalculate_q!, recalculate_F!

	function recalculate_variables!(
		q::Matrix{T},
		s::State{T};
		update_speed_of_sound::Bool=false
	) where T <: AbstractFloat
		s.ρ .= q[1, :]
		s.u .= q[2, :] ./ s.ρ
		s.E .= q[3, :] ./ s.ρ
		s.e .= s.E - ((1/2) .* s.u.^2)
		s.p .= (s.gamma - 1) .* s.ρ .* s.e

		if update_speed_of_sound 
			s.c .= sqrt.(s.gamma .* s.p ./ s.ρ)
		end
	end

	function recalculate_q!(
		q::Matrix{T},
		s::State,
	) where T <: AbstractFloat
		q[1, :] = s.ρ
		q[2, :] = s.ρ .* s.u
		q[3, :] = s.ρ .* s.E
	end

	function recalculate_F!(
		F::Matrix{T}, 
		s::State,
	) where T <: AbstractFloat
		F[1, :] = s.ρ .* s.u
		F[2, :] = (s.ρ .* s.u.^2) .+ s.p
		F[3, :] = s.u .* ((s.ρ .* s.E) .+ s.p)
	end

	function adjust_Fstar!(
		F::Matrix{T}, 
		s::State,
	) where T <: AbstractFloat
		n = length(s.u)

		du_dx_buffer = copy(s.u)
		forward_difference!(s.u, du_dx_buffer, s.h)

		vec = zeros(3,n)
		vec[2, :] .= 1
		vec[3, :] = s.u

		term = s.alpha * s.h^2 .* s.ρ .* abs.(du_dx_buffer) .* du_dx_buffer .* transpose(vec[:, 1:n])

		F -= transpose(term)
		return F
	end

	function forward_difference!(
		input::Vector{T},
		output::Vector{T},
		h::T
	) where T <: AbstractFloat
		for i = 1:length(input)-1
			output[i] = (input[i+1] - input[i]) / h
		end
	end
end

File: src/vector_calculations.rs
use crate::prelude::*;


pub(crate) fn recalculate_q(q: &mut Matrix, s: &State) {
    for i in 0..s.n {
        q[[0, i]] = s.rho[i]
    }

    for i in 0..s.n {
        q[[1, i]] = s.rho[i] * s.u[i]
    }

    for i in 0..s.n {
        q[[2, i]] = s.rho[i] * s.E[i]
    }
}

pub(crate) fn recalculate_f(f: &mut Matrix, s: &State) {
    for i in 0..s.n {
        f[[0, i]] = s.rho[i] * s.u[i]
    }

    for i in 0..s.n {
        f[[1, i]] = (s.rho[i] * s.u[i].powf(2.)) + s.p[i]
    }

    for i in 0..s.n {
        f[[2, i]] = s.u[i] * (s.p[i] + (s.rho[i] * s.E[i]))
    }
}

pub(crate) fn recalculate_variables(q: &Matrix, s: &mut State) {
    for i in 0..s.n {
        s.rho[i] = q[[0, i]]
    }

    for i in 0..s.n {
        s.u[i] = q[[1, i]] / s.rho[i]
    }

    for i in 0..s.n {
        s.E[i] = q[[2, i]] / s.rho[i]
    }

    for i in 0..s.n {
        s.e[i] = s.E[i] - (0.5 * s.u[i].powf(2.));

        if s.E[i] < 0.5 * s.u[i].powf(2.) {
            println!("e will have negative at i = {i}");
        }
    }

    for i in 0..s.n {
        s.p[i] = (s.gamma - 1.) * s.rho[i] * s.e[i]
    }

    for i in 0..s.n {
        s.c[i] = (s.gamma * s.p[i] / s.rho[i]).sqrt();
    }
}
